pkg/awsapi/generate/add_import.go

```diff
@@ -0,0 +1,73 @@
+//go:build ignore
+// +build ignore
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"go/format"
+	"go/token"
+	"io/ioutil"
+	"os"
+	"strconv"
+
+	"github.com/kris-nova/logger"
+
+	"github.com/dave/dst"
+	"github.com/dave/dst/decorator"
+)
+
+func addDotImport(code []byte, packageName string) error {
+	d := decorator.NewDecorator(token.NewFileSet())
+	f, err := d.Parse(code)
+	if err != nil {
+		return err
+	}
+
+	for _, dl := range f.Decls {
+		gd, ok := dl.(*dst.GenDecl)
+		if !ok {
+			logger.Warning("expected type %T; got %T", &dst.GenDecl{}, dl)
+			continue
+		}
+
+		if gd.Tok == token.IMPORT {
+			gd.Specs = append(gd.Specs, &dst.ImportSpec{
+				Name: &dst.Ident{
+					Name: ".",
+				},
+				Path: &dst.BasicLit{Kind: token.STRING, Value: strconv.Quote(packageName)},
+			})
+		}
+	}
+
+	restorer := decorator.NewRestorer()
+	var buf bytes.Buffer
+	if err := restorer.Fprint(&buf, f); err != nil {
+		return fmt.Errorf("error restoring source: %w", err)
+	}
+	src, err := format.Source(buf.Bytes())
+	if err != nil {
+		return fmt.Errorf("error formatting source: %w", src)
+	}
+	if _, err := os.Stdout.Write(src); err != nil {
+		return fmt.Errorf("error writing source to STDOUT: %w", err)
+	}
+	return nil
+}
+
+func main() {
+	if len(os.Args) != 2 {
+		panic(fmt.Sprintf("usage: add_import <package-to-import>"))
+	}
+	source, err := ioutil.ReadAll(os.Stdin)
+	if err != nil {
+		panic(fmt.Errorf("unexpected error reading from STDIN: %w", err))
+	}
+
+	packageName := os.Args[1]
+	if err = addDotImport(source, packageName); err != nil {
+		panic(err)
+	}
+}
```

Comment:
is this really easier than just having the sed :thinking: ?

Reply:
It's by no means easier than sed, but it's more robust IMO. We can also modify this program later to avoid using dot imports by qualifying local types with the package name.

Reply:
>  We can also modify this program later to avoid using dot imports by qualifying local types with the package name.

do we need to do that though :shrug: ?

> It's by no means easier than sed, but it's more robust IMO.

I would disagree, the contents of `addImport` is pretty obscure, on a first pass I struggle to understand what its disagree. The sed is just a bit simpler to read, and much easier to test manually

Reply:
>  it's more robust IMO

I'm trying to understand the pros/cons of each `sed` vs `addImport`. @cPu1 could you expand a bit on this, please? `addDotImport` adds quite a lot of complexity compared to `sed`. What do you see as the issues or limitations of `sed`, and how does `addDotImport ` solve this and perform better?

If we do go with this program, I would suggest adding some tests to ensure that it really is more robust and to make it more maintainable.

Reply:
@nikimanoledaki `sed` is not the right tool to manipulate Go sources. That `sed` snippet can generate programs that have syntax errors.

Here's what's wrong with using that sed snippet:

```bash
$ cat <<EOF | sed "/\"context\".*/a . \"github.com/aws/aws-sdk-go-v2/service/eks\""
// Code generated by ifacemaker; DO NOT EDIT.

package awsapi

import "context"

// CloudFormation provides an interface to the AWS CloudFormation service.
type CloudFormation interface {
 // ...
}

EOF

// Code generated by ifacemaker; DO NOT EDIT.

package awsapi

import "context"
. "github.com/aws/aws-sdk-go-v2/service/eks"

// CloudFormation provides an interface to the AWS CloudFormation service.
type CloudFormation interface {
 // ...
}


$ cat <<EOF | sed "/\"context\".*/a . \"github.com/aws/aws-sdk-go-v2/service/eks\""
// Code generated by ifacemaker; DO NOT EDIT.
// Copyright 2020 Weaveworks.
// The awsapi package provides an interface to AWS services. The "context" package is used to carry deadlines and
// cancellation signals.

package awsapi

import (
    "context"
)

// CloudFormation provides an interface to the AWS CloudFormation service.
type CloudFormation interface {
 // ...
}

EOF

// Code generated by ifacemaker; DO NOT EDIT.
// Copyright 2020 Weaveworks.
// The awsapi package provides an interface to AWS services. The "context" package is used to carry deadlines and
. "github.com/aws/aws-sdk-go-v2/service/eks"
// cancellation signals.

package awsapi

import (
    "context"
. "github.com/aws/aws-sdk-go-v2/service/eks"
)

// CloudFormation provides an interface to the AWS CloudFormation service.
type CloudFormation interface {
 // ...
}


$ cat <<EOF | sed "/\"context\".*/a . \"github.com/aws/aws-sdk-go-v2/service/eks\""
package awsapi

import (
    "fmt"
)

// CloudFormation provides an interface to the AWS CloudFormation service.
type CloudFormation interface {
 // ...
}

EOF

package awsapi

import (
    "fmt"
)

// CloudFormation provides an interface to the AWS CloudFormation service.
type CloudFormation interface {
 // ...
}

```



As for the maintenance of the `add_import.go` program, we have an issue for that: https://github.com/weaveworks/eksctl/issues/4994